text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,888 @@
+`Logging.of(construct)` API:
+
+Instead of:
+
+```ts
+construct.node.addWarning('my warning');
+```
+
+Use:
+
+```ts
+import { Logging } from '@aws-cdk/core';
+
+Logging.of(construct).addWarning('my warning');
+```
+
+See [examples](https://github.com/aws/aws-cdk/pull/9056/commits/e4dff913d486592b1899182e9a928765553654fa).
+
+# Motivation
+
+There are various motivations for this change:
+
+1. Removal of a redundant layer
+2. User confusion caused by multiple `Construct` types
+3. Inability to compose AWS CDK constructs into other domains
+4. Non-intuitive dependency requirement on `constructs`
+
+## 1. Redundant layer
+
+The current compatibility layer does not have any logic in it. It is pure glue
+introduced in order to avoid breaking v1.x users. As we release v2.0 we are able
+to clean up this layer, and with it, improve maintainability and code hygiene.
+
+## 2. Multiple `Construct` types
+
+The current situation is error-prone since we have two `Construct` classes in
+the type closure. For example, when a developer types `"Construct"` and uses
+VSCode to _automatically add import statements_, the IDE will actually add an
+import for `constructs.Construct`. If they define a custom construct class which
+extends this type instead of the `core.Construct` type, it won't be possible to
+pass an instance of this class as a scope to AWS CDK constructs such as `Stack`
+for example.
+
+## 3. Composability with other domains
+
+The main motivation for this change is to enable composition of AWS CDK
+constructs with constructs from other domains.
+
+It is currently impossible to define AWS CDK constructs within a non-AWS-CDK
+construct scope.
+
+For example, consider the
+[Terrastack](https://github.com/TerraStackIO/terrastack) project or a similar
+one, which uses CDK constructs to define stacks through Terraform. Say we want to
+use an AWS CDK L2 inside a Terraform stack construct:
+
+```ts
+const stack = new terrastack.Stack(...);
+
+// COMPILATION ERROR: `this` is not a `cdk.Construct`.
+new s3.Bucket(this, 'my-bucket');
+```
+
+Being able to create construct compositions from multiple domains is a powerful
+direction for the CDK ecosystem, and this change is required in order to enable
+these use case.
+
+## 4. Non-intuitive dependency requirement
+
+As we transition to [monolithic packaging] as part of v2.x, CDK users will have
+to take a _peer dependency_ on both the CDK library (`aws-cdk-lib`) and
+`constructs`.
+
+> Currently, the AWS CDK also takes `constructs` as a normal dependency (similar
+> to all dependencies), but this is about to change with mono-cdk.
+
+The reason `constructs` will also be required (whether we leave the
+compatibility layer or not) is due to the fact that all CDK constructs
+eventually extend the base `constructs.Construct` class. This means that this
+type is part of their public API and therefore a peer dependency is required
+(otherwise, there could be incompatible copies of `Construct` in the node
+runtime).
```

Comment:
added clarification:

The removal of the compatibility layer means that now anyone who uses the AWS
CDK will need to explicitly use the `constructs.Construct` type (even for
trivial apps), and therefore it would "make sense" for them to take a dependency
on the `constructs` library.